### PR TITLE
Управление балансом

### PR DESCRIPTION
**Pull request by NikitaGoroshko7**
https://trello.com/c/LvT0FXaW/40-%D1%83%D0%BF%D1%80%D0%B0%D0%B2%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5-%D0%B1%D0%B0%D0%BB%D0%B0%D0%BD%D1%81%D0%BE%D0%BC

Has been added PayController for payment
Has been added PayViewModel
Has been added actions Payment for PayController
Has been added Views for PayController
Has been added CreaditCard style and script for Payment View
Has been edited modsl windows for Profile and Settings
Has been added modal window for Balance

### CODE REVIEW

File: CrowdFundingApplication/Controllers/PayController.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public PayController(UserManager<User> userManager)
+    {
+        _userManager = userManager;
+    }
+
+    [HttpGet]
+    public IActionResult Payment([FromQuery(Name = "userId")] string userId, [FromQuery(Name = "Sum")] int Sum)
+    {
+        ViewBag.Sum = Sum;//seting sum in ViewBag for display 
+        return userId is null ? RedirectToAction("Balance","Profile") : View(); //check if user entered in url ~/Pay/Payment
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/PayController.cs]: spase

File: CrowdFundingApplication/Views/Pay/Payment.cshtml [hunk 2]
```diff
@@ -0,0 +1,87 @@
+﻿@model CrowdFundingApplication.Models.PayViewModel
+
+@{
+    Layout = null;
```
Comment (GalayM) [inline CrowdFundingApplication/Views/Pay/Payment.cshtml]: Use Layout!

File: CrowdFundingApplication/Controllers/PayController.cs (hunk 1 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/PayController.cs]: has been edited

File: CrowdFundingApplication/appsettings.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "DefaultConnection": "Server=CMDB-80646;Database=CrowdfundingApp;Trusted_Connection=True;MultipleActiveResultSets=true"
+    "DefaultConnection": "Server=DESKTOP-0RUJG5Q\\SQLEXPRESS;Database=CrowdfundingApp;Trusted_Connection=True;MultipleActiveResultSets=true"
```
Comment (GalayM) [inline CrowdFundingApplication/appsettings.json]: Instead, DESKTOP-0RUJG5Q\\SQLEXPRESS uses alias.

File: CrowdFundingApplication/wwwroot/js/CreaditCard.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
+﻿/* Card Number Spacing */
+$('#card-number').on('keypress change blur', function () {
```
Comment (GalayM) [inline CrowdFundingApplication/wwwroot/js/CreaditCard.js:2]: I think it's better to use a ready-made library for this.

File: CrowdFundingApplication/wwwroot/js/CreaditCard.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+$('#card-exp').on('input', function () {
+    var curLength = $(this).val().length;
+    if (curLength === 2) {
+        var newInput = $(this).val();
+        newInput += '/';
+        $(this).val(newInput);
+    }
+});
+
+/* Validation for creadit card*/
```
Comment (GalayM) [inline CrowdFundingApplication/wwwroot/js/CreaditCard.js:24]: I think it's better to use a ready-made library for this.

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 6]
```diff
@@ -79,6 +79,13 @@ public async Task<IActionResult> Balance()
         return View(GetUserViewModel(user));
     }
 
+    [HttpPost]
+    public async Task<IActionResult> Balance(int Sum)
+    {   
+        var user = await _userManager.FindByNameAsync(User.Identity.Name);
+        return Sum != 0 ? RedirectToAction("Payment", "Pay", new { userId = user.Id, Sum = Sum }) : View();
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: What happens if the sum is negative?

Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/appsettings.json (hunk 3 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/appsettings.json]: Has been edited

File: CrowdFundingApplication/Views/Pay/Payment.cshtml (hunk 2 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Views/Pay/Payment.cshtml]: Has been edited

File: CrowdFundingApplication/Controllers/PayController.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    public PayController(UserManager<User> userManager)
+    {
+        _userManager = userManager;
+    }
+
+    [HttpGet]
+    public IActionResult Payment([FromQuery(Name = "userId")] string userId, [FromQuery(Name = "Sum")] int Sum)
+    {
+        var model = new PayViewModel { Sum = Sum.ToString() };//seting sum in ViewBag for display
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/PayController.cs]: why string?

File: CrowdFundingApplication/Controllers/PayController.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+        return userId is null ? RedirectToAction("Balance","Profile") : View(model); //check if user entered in url ~/Pay/Payment
+    }
+
+    [HttpPost]
+    public async Task<IActionResult> Payment(PayViewModel model)
+    {
+        if (ModelState.IsValid)
+        {
+            var getModelSum = model.Sum.Split(" ");//geting sum from model
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/PayController.cs]: What is this?

File: CrowdFundingApplication/appsettings.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "DefaultConnection": "Server=CMDB-80646;Database=CrowdfundingApp;Trusted_Connection=True;MultipleActiveResultSets=true"
+    "DefaultConnection": "Server=.\\SQLEXPRESS;Database=CrowdfundingApp;Trusted_Connection=True;MultipleActiveResultSets=true"
```
Comment (GalayM) [inline CrowdFundingApplication/appsettings.json]: .\\SQLEXPRESS -> .

File: CrowdFundingApplication/Controllers/PayController.cs (hunk 7 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/PayController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/PayController.cs (hunk 8 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/PayController.cs]: Has been edited

File: CrowdFundingApplication/appsettings.json (hunk 9 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/appsettings.json]: Has been edited

File: CrowdFundingApplication/Controllers/PayController.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        return userId is null ? RedirectToAction("Balance","Profile") : View(model); //check if user entered in url ~/Pay/Payment
+    }
+
+    [HttpPost]
+    public async Task<IActionResult> Payment(PayViewModel model)
+    {
+        if (ModelState.IsValid)
+        {
+            var user = await _userManager.FindByNameAsync(User.Identity.Name);
+            user.Balance += Convert.ToDecimal(model.Sum);//update balance for user
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/PayController.cs]: model.Sum to decimal type? in all cases

File: CrowdFundingApplication/Views/Pay/Payment.cshtml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        <div class="box">
+
+            <div class="mdl">
+                <div class="circles">
+                    <div class="circle circle-1"></div>
+                    <div class="circle circle-2"></div>
+                </div>
+
+                <label style="width:400px;margin-left: 70px;font-size: 24px; color: white;">Сумма пополнения: </label>
+                <input asp-for="Sum" name="Sum" value="@Model.Sum" readonly style="width:400px;margin-left: 70px;margin-bottom: 20px;font-size: 24px;"/>
```
Comment (GalayM) [inline CrowdFundingApplication/Views/Pay/Payment.cshtml]: styles to css files, in all cases

File: CrowdFundingApplication/Controllers/PayController.cs (hunk 10 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/PayController.cs]: Has been edited

File: CrowdFundingApplication/Views/Pay/Payment.cshtml (hunk 11 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Views/Pay/Payment.cshtml]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 12]
```diff
@@ -79,6 +79,22 @@ public async Task<IActionResult> Balance()
         return View(GetUserViewModel(user));
     }
 
+    [HttpPost]
+    public async Task<IActionResult> Balance(int Sum)
+    {   
+        var user = await _userManager.FindByNameAsync(User.Identity.Name);
+        if(Sum != 0 && Sum > 0)
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: Why this Sum != 0 condition?

File: CrowdFundingApplication/Controllers/ProfileController.cs [hunk 13]
```diff
@@ -79,6 +79,22 @@ public async Task<IActionResult> Balance()
...
+        {
+            return RedirectToAction("Payment", "Pay", new { userId = user.Id, Sum = Sum });
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/ProfileController.cs]: Payment action gets the specific type

File: CrowdFundingApplication/Controllers/PayController.cs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+{
+    private readonly UserManager<User> _userManager;
+
+    public PayController(UserManager<User> userManager)
+    {
+        _userManager = userManager;
+    }
+
+    [HttpGet]
+    public IActionResult Payment([FromQuery(Name = "userId")] string userId, [FromQuery(Name = "Sum")] int Sum)
```
Comment (GalayM) [inline CrowdFundingApplication/Controllers/PayController.cs]: Sum -> sum

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 12 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/ProfileController.cs (hunk 13 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/ProfileController.cs]: Has been edited

File: CrowdFundingApplication/Controllers/PayController.cs (hunk 14 above)
Comment (NikitaGoroshko7) [reply CrowdFundingApplication/Controllers/PayController.cs]: Has been edited

**Review by GalayM — APPROVED** [review]

**Pull request merged**